File: renderer/flutter/lib/controller/web_view.dart
//
// Tencent is pleased to support the open source community by making
// Hippy available.
//
// Copyright (C) 2019 THL A29 Limited, a Tencent company.
// All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import '../common.dart';
import '../controller.dart';
import '../render.dart';
import '../style.dart';
import '../viewmodel.dart';
import '../widget.dart';

class WebViewViewController extends BaseViewController<WebViewModel> {
  static const String kClassName = "WebView";

  static const String kUserAgent = "userAgent";
  static const String kMethod = "method";
  static const String kOnLoad = "onLoad";
  static const String kOnLoadStart = "onLoadStart";
  static const String kOnLoadEnd = "onLoadEnd";
  static const String kOnError = "onError";

  @override
  WebViewModel createRenderViewModel(RenderNode node, RenderContext context) {
    return WebViewModel(node.id, node.rootId, node.name, context);
  }

  @override
  Widget createWidget(BuildContext context, WebViewModel viewModel) {
    return WebViewWidget(viewModel);
  }

  @override
  Map<String, ControllerMethodProp> get extendRegisteredMethodProp {
    var extraMap = <String, ControllerMethodProp>{};
    extraMap[NodeProps.kSource] = ControllerMethodProp(setSource, null);
    extraMap[kUserAgent] = ControllerMethodProp(setUserAgent, '');
    extraMap[kMethod] = ControllerMethodProp(setMethod, '');
    extraMap[kOnLoad] = ControllerMethodProp(setOnLoad, false);
    extraMap[kOnLoadStart] = ControllerMethodProp(setOnLoadStart, false);
    extraMap[kOnLoadEnd] = ControllerMethodProp(setOnLoadEnd, false);
    extraMap[kOnError] = ControllerMethodProp(setOnError, false);

    return extraMap;
  }

  @ControllerProps(NodeProps.kSource)
  void setSource(WebViewModel renderViewModel, VoltronMap source) {
    String src = source.get('uri');
    if (src != renderViewModel.src) {
      renderViewModel.src = src;
    }
  }

  @ControllerProps(kUserAgent)
  void setUserAgent(WebViewModel renderViewModel, String ua) {
    if (ua != renderViewModel.userAgent) {
      renderViewModel.userAgent = ua;
    }
  }

  @ControllerProps(kMethod)
  void setMethod(WebViewModel renderViewModel, String method) {
    renderViewModel.method = method;
  }

  @ControllerProps(kOnLoadStart)
  void setOnLoadStart(WebViewModel renderViewModel, bool flag) {
    renderViewModel.onLoadStartEnable = flag;
  }

  @ControllerProps(kOnError)
  void setOnError(WebViewModel renderViewModel, bool flag) {
    renderViewModel.onErrorEnable = flag;
  }

  @ControllerProps(kOnLoadEnd)
  void setOnLoadEnd(WebViewModel renderViewModel, bool flag) {
    renderViewModel.onLoadEndEnable = flag;
  }

  @ControllerProps(kOnLoad)
  void setOnLoad(WebViewModel renderViewModel, bool flag) {
    renderViewModel.onLoadEnable = flag;
  }

  @override
  String get name => kClassName;
}

File: packages/hippy-web-renderer/src/component/web-view.ts
/*
 * Tencent is pleased to support the open source community by making
 * Hippy available.
 *
 * Copyright (C) 2022 THL A29  Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {  NodeProps, HippyBaseView, InnerNodeTag  } from '../types';
import { HippyWebView } from './hippy-web-view';

export class WebView extends HippyWebView<HTMLIFrameElement> {
  private isMounted = false;
  public constructor(context, id, pId) {
    super(context, id, pId);
    this.dom = document.createElement('iframe');
    this.tagName = InnerNodeTag.WEB_VIEW;
    this.dom!.onload = this.onLoadEnd.bind(this);
  }

  public set source(value: { uri: string }) {
    this.props[NodeProps.SOURCE] = value;
    if (this.isMounted) {
      this.onLoadStart(this.props[NodeProps.SOURCE]);
      this.dom!.src = value.uri ?? '';
      this.onLoad(this.props[NodeProps.SOURCE]);
    }
  }

  public get source() {
    return this.props[NodeProps.SOURCE];
  }

  public set userAgent(value: string) {
    this.props[NodeProps.USER_AGENT] = value;
  }

  public get userAgent() {
    return this.props[NodeProps.USER_AGENT];
  }

  public set method(value: string) {
    this.props[NodeProps.METHOD] = value;
  }

  public get method() {
    return this.props[NodeProps.METHOD];
  }

  public  onLoadStart(value: { uri: string }) {
    this.props[NodeProps.ON_LOAD_START] && this.context.sendUiEvent(this.id, NodeProps.ON_LOAD_START, value);
  }

  public  onLoad(value: { uri: string }) {
    this.props[NodeProps.ON_LOAD] && this.context.sendUiEvent(this.id, NodeProps.ON_LOAD, value);
  }

  public  onLoadEnd() {
    this.props[NodeProps.ON_LOAD_END] && this.context.sendUiEvent(
      this.id,
      NodeProps.ON_LOAD_END, this.props[NodeProps.SOURCE],
    );
  }

  public async beforeMount(parent: HippyBaseView, position: number) {
    await super.beforeMount(parent, position);
    if (this.props[NodeProps.SOURCE].uri) this.dom!.src = this.props[NodeProps.SOURCE].uri ?? '';
    if (this.dom!.src) {
      this.onLoadStart(this.props[NodeProps.SOURCE]);
    }
  }

  public mounted() {
    this.isMounted = true;
    if (this.dom!.src) {
      this.onLoad(this.props[NodeProps.SOURCE]);
    }
  }

  public set style(value: { uri: string }) {
    this.props[NodeProps.STYLE] = value;
  }

  public get style() {
    return this.props[NodeProps.STYLE];
  }
}
